### PR TITLE
CI: Add Wasmtime based Wasm benchmark automation

### PR DESCRIPTION
**Pull request by sergejparity**
Addresses https://github.com/paritytech/wasmi/issues/443

Pipeline updated and now is divided in two stages, `benchmarks` (perform benchmarking) and `report` parses results, generate report and post it into PR.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
## CRITERION BENCHMARKS ## 

 |BENCHMARK|MASTER|PR|DIFF|MASTER(WASM)|PR(WASM)|DIFF(WASM)|
 |---|---:|---:|---|---:|---:|---|
 | `execute_count_until`  | 2.2426 ms  | 2.4737 ms  | 12.9988  % | 8.9714 ms  | 9.3121 ms  | 4.0678 % |
| `execute_factorial_iterative`  | 1.5261 µs  | 1.6648 µs  | 9.0737  % | 3.4226 µs  | 3.4155 µs  | -0.1011 % |
| `execute_factorial_recursive`  | 2.0757 µs  | 2.0634 µs  | -0.6236  % | 4.1976 µs  | 4.1608 µs  | -0.9863 % |
| `execute_fib_iterative`  | 9.8916 ms  | 11.2896 ms  | 13.9383  % | 15.5359 ms  | 16.0466 ms  | 3.4030 % |
| `execute_fib_recursive`  | 18.3662 ms  | 18.4565 ms  | 1.8211  % | 38.4940 ms  | 38.7383 ms  | 0.4658 % |
| `execute_global_bump`  | 3.5386 ms  | 3.4510 ms  | -2.5589  % | 12.9254 ms  | 12.8407 ms  | -0.6683 % |
| `execute_host_calls`  | 33.3589 µs  | 33.0163 µs  | -0.5980  % | 114.4963 µs  | 109.6648 µs  | -4.2408 % |
| `execute_memory_fill`  | 8.1961 ms  | 9.5507 ms  | 16.4551  % | 14.9306 ms  | 15.1120 ms  | 0.8858 % |
| `execute_memory_sum`  | 7.9422 ms  | 9.0321 ms  | 13.6214  % | 15.3984 ms  | 15.3832 ms  | -0.0395 % |
| `execute_memory_vec_add`  | 14.8578 ms  | 15.8455 ms  | 6.6518  % | 30.6651 ms  | 30.8989 ms  | 0.7619 % |
| `execute_recursive_is_even`  | 2.8290 ms  | 2.6875 ms  | -4.4542  % | 6.6252 ms  | 6.4608 ms  | -2.4676 % |
| `execute_recursive_ok`  | 378.7953 µs  | 396.0410 µs  | 4.9397  % | 906.8774 µs  | 893.2604 µs  | -1.4101 % |
| `execute_recursive_scan`  | 490.6378 µs  | 465.4433 µs  | -3.7205  % | 1.2876 ms  | 1.2652 ms  | -1.6241 % |
| `execute_recursive_trap`  | 38.4758 µs  | 38.0965 µs  | -1.2980  % | 92.9290 µs  | 92.9484 µs  | -0.0501 % |
| `execute_regex_redux`  | 3.3814 ms  | 3.3735 ms  | -0.2180  % | 5.1031 ms  | 5.0791 ms  | -0.5336 % |
| `execute_rev_complement`  | 3.0258 ms  | 3.1499 ms  | 4.2607  % | 4.9786 ms  | 4.9685 ms  | -0.2244 % |
| `execute_tiny_keccak`  | 2.6299 ms  | 2.6983 ms  | 2.6366  % | 4.2362 ms  | 4.2302 ms  | -0.1986 % |
| `execute_trunc_f2i`  | 3.5213 ms  | 3.8386 ms  | 9.0142  % | 8.8933 ms  | 8.9013 ms  | 0.0910 % |
| `instantiate_wasm_kernel`  | 77.4873 µs  | 77.8239 µs  | 0.8983  % | 124.4840 µs  | 158.3217 µs  | 27.1820 % |
| `translate_wasm_kernel`  | 6.1144 ms  | 6.0945 ms  | -0.4291  % | 12.8741 ms  | 12.7245 ms  | -0.7712 % |


 [Link to pipeline](https://gitlab.parity.io/parity/mirrors/wasmi/-/jobs/1894010)

**Comment by Robbepop** [comment]
We could replace `(WASM)` in the table header with the [official Wasm icon](https://upload.wikimedia.org/wikipedia/commons/thumb/1/1f/WebAssembly_Logo.svg/1200px-WebAssembly_Logo.svg.png) to spare to space.

File: scripts/ci/benchmarks-report.sh [hunk 1]
```diff
@@ -38,9 +69,19 @@ curl -X ${REQUEST_TYPE} ${PR_COMMENTS_URL} \
...
 \"body\": \
-\"## CRITERION BENCHMARKS ## \n\n \
-|BENCHMARK|MASTER|PR|Diff|\n \
-|---|---:|---:|---|\n \
-${RESULT}\n\n \
+\"## BENCHMARKS ## \n\n \
+<table> \
+<thead> \
+<tr> \
+<th><\/th><th colspan=3>CARGO BENCH<\/th><th colspan=3>WASMTIME<\/th> \
```
[the remark this answers is not in the document]
Comment (sergejparity) [reply scripts/ci/benchmarks-report.sh:89]: Yes. Sure.
I gave up trying to format table using pure Markdown. And switched to HTML as it is much more flexible

[the remark this answers is not in the document]
Comment (Robbepop) [reply scripts/ci/benchmarks-report.sh:89]: HTML is probably the better choice here.

**Comment by athei** [comment]
Nice. The numbers are without wasm-opt, right?

**Comment by Robbepop** [comment]
> Nice. The numbers are without wasm-opt, right?

We are using `cargo-wasi` to build `wasmi` which uses `wasm-opt` post-optimziation hook out of the box. So I have to frustrate you to say that those numbers are already how good it gets. The overhead from running via Wasmtime is at least ~50% compared to native and sometimes even much much more than that.

**Comment by athei** [comment]
I am asking because it would be very interesting to compare different optimization levels.

**Comment by Robbepop** [comment]
> I am asking because it would be very interesting to compare different optimization levels.

https://bytecodealliance.github.io/cargo-wasi/wasm-opt.html#selected-optimization-level

**Comment by athei** [comment]
What I am suggesting is that the CI prints both numbers: Unoptimized vs.`O3` for example. I remember there was some crazy stuff going on like unoptimized being faster. We need to catch this.

**Review by athei — APPROVED** [review]
Sure :)

**Comment by Robbepop** [comment]
@sergejparity is this done?

If "yes", could you please make the benchmark names `monospace` again?

**Comment by sergejparity** [comment]
> If "yes", could you please make the benchmark names `monospace` again before we merge?

Done

**Comment by Robbepop** [comment]
@sergejparity  CI is red because `Criterion` is using `default-features` in `master`, right?

**Comment by sergejparity** [comment]
> The only thing that looks odd to me is that the NATIVE side is extremely noisy ...

Me too. Benchmarks gives unstable results. The same picture I observe performing them locally. Probably some tuning is needed, increase iterations count.

**Comment by sergejparity** [comment]
> @sergejparity CI is red because `Criterion` is using `default-features` in `master`, right?

Yes, right.

**Comment by Robbepop** [comment]
> 

It isn't as unstable on my system locally though. Also the Wasmtime benchmarks are less unstable but the `instantiate` benchmark is always faster in the PR branch strangely.

**Review by Robbepop — APPROVED** [review]
LGTM!

Thanks a lot for this important enhancement of our benchmarking CI @sergejparity !
The table results display really well and look fantastic. :rocket:
I already have a few follow-ups for future extensions in the back of my head. :blush:

**Pull request merged**